File: 4-median-of-two-sorted-arrays.rs
// Median of Two Sorted Arrays
// #url: https://leetcode.com/problems/median-of-two-sorted-arrays/
// #hard
// #binary_search
//
// Example:
// nums1 = [1, 3]
// nums2 = [2]
// The median is 2.0
//
// Example:
// nums1 = [1, 2]
// nums2 = [3, 4]
// The median is (2 + 3)/2 = 2.5
//
// Questions to ask:
//   1. Can be empty?
//   2. Runtime complexity?
//      The overall run time complexity should be O(log (m+n)).
//
// Algo:
//   Step 1:
//     [1, 3, 5], [2, 4, 6, 8]
//     l1 = 4
//     l2 = 3
//     imin = 0
//     imax = 4
//     half_len = (4 + 3 + 1) / 2 = 4
//
//   Step 2 (while iter 1):
//     i = (0 + 4) / 2 = 2
//     j = 4 - 2 = 2
//     [2, 4, (6), 8], [1, 3, (5)],
//             i               j
//     imin = 1 + 1 = 2
//
//   Step 3 (while iter 1):
//     i = (2 + 3) / 2 = 2
//     j = 4 - 2 = 2
//     [1, 3, (5)], [2, 4, (6), 8]
//             i           j
//     max_left = max(3, 4) = 4

use std::cmp;

fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
  let (arr1, arr2) = if nums1.len() > nums2.len() {
    (nums1, nums2)
  } else {
    (nums2, nums1)
  };
  let arr1_len = arr1.len();
  let arr2_len = arr2.len();
  let mut imin = 0;
  let mut imax = arr1_len;
  let half_len = (arr1_len + arr2_len + 1) / 2;

  while imin <= imax {
    let i = (imin + imax) / 2;
    let j = half_len - i;

    println!("i: {:?}, j: {:?}, imax: {:?}", i, j, imax);

    if i < imax && arr2[j - 1] > arr1[i] {
      imin = i + 1;
    } else if i > imin && arr1[i - 1] > arr2[j] {
      imax = i - 1;
    } else {
      let max_left: i32;
      if i == 0 {
        max_left = arr2[j - 1];
      } else if j == 0 {
        max_left = arr1[i - 1];
      } else {
        max_left = cmp::max(arr1[i - 1], arr2[j - 1]);
      }


      if (arr1_len + arr2_len) % 2 == 1 {
        return max_left as f64;
      }

      let min_right: i32;
      if i == arr1_len {
        min_right = arr2[j];
      } else if j == arr2_len {
        min_right = arr1[i];
      } else {
        min_right = cmp::min(arr2[j], arr1[i]);
      }

      return ((max_left + min_right) as f64) / 2.0;
    }
  }

  0.0
}

fn main() {
  println!(
    "{:?}",
    find_median_sorted_arrays(vec![1, 3, 5], vec![2, 4, 6, 8])
  );
}

#[cfg(test)]
mod tests {
  // Note this useful idiom: importing names from outer (for mod tests) scope.
  use super::*;

  #[test]
  fn test_median_n1n2() {
    assert_eq!(find_median_sorted_arrays(vec![1, 2], vec![3, 4]), 2.5);
  }

  #[test]
  fn test_median_n2n1() {
    assert_eq!(find_median_sorted_arrays(vec![3, 4], vec![1, 2]), 2.5);
  }

  #[test]
  fn test_median_n1n2_overlap() {
    assert_eq!(
      find_median_sorted_arrays(vec![1, 3, 5], vec![2, 4, 6, 8]),
      4.0
    );
  }
}

File: python/leet/4-median-of-two-sorted-arrays.py
# Median of Two Sorted Arrays
# https://leetcode.com/problems/median-of-two-sorted-arrays/
# hard
#
# Time:  O(log (m+n))
#
# Solution:
# Binary Search

def findMedianSortedArrays(nums1: [int], nums2: [int]) -> float:
    posn1 = posn2 = cur = prev = 0
    mid = (len(nums1) + len(nums2)) / 2.0

    while mid >= 0.0:
        item1 = nums1[posn1] if posn1 < len(nums1) else float("inf")
        item2 = nums2[posn2] if posn2 < len(nums2) else float("inf")

        prev = cur
        mid -= 1.0

        if item1 <= item2:
            cur = posn1
            posn1 += 1
        else:
            cur = posn2 + len(nums1)
            posn2 += 1

    item1 = nums2[cur-len(nums1)] if cur >= len(nums1) else nums1[cur]
    item2 = nums2[prev-len(nums1)] if prev >= len(nums1) else nums1[prev]
    return (item1 + item2) / 2.0 if (len(nums1) + len(nums2)) % 2.0 == 0 else float(item1)


print(findMedianSortedArrays([1,3], [2]), 2.0)
print(findMedianSortedArrays([1,2], [3,4]), 2.5)
print(findMedianSortedArrays([1,3], [2,4]), 2.5)
print(findMedianSortedArrays([1,5], [2,10]), 3.5)
print(findMedianSortedArrays([1], []), 1.0)
print(findMedianSortedArrays([], [2]), 2.0)

#
# 1,    3
#    2,
# mid = 2
# cur = 2
# i0 = 0
# i1 = 0

# [1,2] [3,4]
#      ^ ^
# posn1 = 0   | 1   | 2   | 2     | | | | | | | | | | |
# posn2 = 0   | 0   | 0   | 1     | | | | | | | | | | |
# mid   = 2.0 | 1.0 | 0.0 | -1.0  | | | | | | | | | | |
# cur   = 0   | 0   | 1   | 2     | | | | | | | | | | |
# prev  = 0   | 0   | 0   | 1     | | | | | | | | | | |
